### PR TITLE
AP_Avoidance/AP_Proximity: simplification and efficiency improvements

### PR DESCRIPTION
**Pull request by rmackay9**
This PR resolves a few issues that affect Copter's BendyRuler object avoidance:

- Object Database and Proximity libraries lose some unnecessary const on arguments passed by value
- Object Database's object's radii are calculated based on their distance from the vehicle (and beam width) rather than based on their direction compared to the vehicle's heading.  Resolves https://github.com/ArduPilot/ardupilot/issues/12661 and makes the circles around objects on the map look more like the users would expect.
- Object Database stores locations as offsets from the EKF origin (in meters) instead of as locations.  This saves on memory (because it uses Vector2f) and CPU (because there are less conversions between locations and offsets-from-origin)
- AP_Proximity's Lightware SF40c driver combines 7 readings together (choosing the shortest) before sending to the Object Database to improve efficiency (both CPU and memory).  7deg is accurate enough and was chosen because normally the sensor sends packets holding 21 readings.
- AP_Proximity library gets simplified handling of ignore zones to improve efficiency.  Previously it would create up to 12 sectors between ignore zones, now it always uses 8 and simply ignores readings if they fall within the ignore zone.
- BendyRuler updates at 1hz instead of 10hz.  This gives enough time for the Copter position controller to get the vehicle moving before another new target arrives.
- Object Database is available regardless of the HAL_MINIMIZE_FEATURES setting.  Removing the database when hal-minimized-features was set was a half measure and is no longer required because the entire PathPlanning class (including the database) are not compiled in when hal-minimize-featuers is set on Copter.  On Rover PathPlanning including the database is always available because it fits within 1MB

This has been tested on a real vehicle using a Cube Orange and SF40c Lidar.

bendyruler-sf40c

### CODE REVIEW

**Comment by peterbarker** [comment]
Why remove const in the cpp files?  They're local variables - and we encourage `const` whereever possible in local variables.  They don't need to be const in the headers.

**Comment by rmackay9** [comment]
@peterbarker, I'm definitely in favour of us making all the arguments const and making the function headers (or whatever we call them) twice as long.  That's a change in our style .. not a fan!

**Comment by peterbarker** [comment]
On Tue, 10 Dec 2019, Randy Mackay wrote:

> @peterbarker, I'm definitely in favour of us making all the arguments const and making the function headers (or whatever we call them) twice as long.

I just got through saying that it shouldn't be done in the headers.  i.e. 
should be in the definitions but not in the declarations.

> That's a change in our style .. not a fan!

I'm a strong fan of having them in there.  They stop bugs happening at 
compile-time and that is not a bad thing.

I'm not going out of my way to change everything as that's pointless code 
churn.

**Comment by rmackay9** [comment]
@peterbarker I hear what you're saying about just putting it in the definitions.. I guess this means just the cpp and not the .h?  But then they don't match which doesn't seem good.

I really wonder how many bugs have been avoided with the addition of the "const" within the function let alone within the arguments.  I definitely think it makes sense to add the const when we're passing by reference.. but not when we're passing by value.

**Comment by rmackay9** [comment]
rebased on master..

**Comment by tridge** [comment]
looks like this enables bendy ruler on all boards? what is the flash cost on Pixhawk1-1M ?

File: libraries/AC_Avoidance/AP_OADatabase.cpp [hunk 1]
```diff
@@ -418,12 +373,15 @@ void AP_OADatabase::send_adsb_vehicle(mavlink_channel_t chan, uint16_t interval_
...
+
         mavlink_msg_adsb_vehicle_send(chan,
             idx,
-            _database.items[idx].loc.lat,
-            _database.items[idx].loc.lng,
+            item_loc.lat,
+            item_loc.lng,
             0,                          // altitude_type
-            _database.items[idx].loc.alt,
+            0,                          // altitude is always zero
```
Comment (tridge) [inline libraries/AC_Avoidance/AP_OADatabase.cpp]: should be our alt above sea level?

**Comment by magicrub** [comment]
I'd like to look this over carefully before merging.

**Comment by rmackay9** [comment]
@magicrub, great, thanks.  Any feedback is greatly appreciated!

File: libraries/AC_Avoidance/AP_OADatabase.cpp [hunk 2]
```diff
@@ -101,45 +127,22 @@ void AP_OADatabase::update()
...
-    } else if (distance < 5 && (angle <= (360-90) || angle >= 90)) {
-        // kinda far and behind us
-        importance = AP_OADatabase::OA_DbItemImportance::Low;
+    // ignore objects that are far away
+    if ((_dist_max > 0.0f) && (distance > _dist_max)) {
+        return;
     }
 
-    const OA_DbItem item = {loc, timestamp_ms, 0, 0, importance};
+    const OA_DbItem item = {pos, timestamp_ms, MAX(_radius_min, distance * dist_to_radius_scalar), 0, AP_OADatabase::OA_DbItemImportance::Normal};
```
Comment (magicrub) [inline libraries/AC_Avoidance/AP_OADatabase.cpp]: forcing everything to importance::Normal will defeat the whole GCS throttling scheme. Setting the Imporantance is still needed, maybe it's just the algorithm that assigns it that needs to change?

Comment (rmackay9) [reply libraries/AC_Avoidance/AP_OADatabase.cpp]: @magicrub,

Yes, I think if we want to try and reduce the number of objects we send to the GCS then improving the algorithm is a good idea and I guess that's what I was trying to do.  While trying to get Bendy Ruler working for Copter I found the display of the objects on the GCS was so unintelligible ([see pic in this issue](https://github.com/ArduPilot/ardupilot/issues/12661)) that I figured it was best to just simplify it down to a point where it worked and then we can re-build from there.

I didn't think that using the vehicle' heading (especially for Copter) was a good idea so what I've done is add the OA_DB_DIST_MAX parameter so the user can specify the maximum distance from the vehicle for any new objects entering the database.  This reduces the number of objects overall both for sending to the GCS but also for the path planning done in BendyRuler.

**Comment by rmackay9** [comment]
I've successfully flown this on a Copter now and I'll produce a video over the next week or so.

File: libraries/AC_Avoidance/AP_OADatabase.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AC_Avoidance/AP_OADatabase.cpp]: Thanks, updated so it sends the adsb messages using the vehicle's current altitude.

**Comment by rmackay9** [comment]
fixed the adsb-send altitude issue and rebased on master.  I plan to merge once it's passing travis.

**Pull request merged**